src/player.h

```diff
@@ -1718,6 +1722,9 @@ class player : public Character
         /** Subset of learned recipes. Needs to be mutable for lazy initialization. */
         mutable pimpl<recipe_subset> learned_recipes;
 
+        /** Recipes that aren't shown in the crafting menu. */
+        mutable pimpl<recipe_subset> hidden_recipes;
```

Comment:
> I've tried to do it with std::unordered_set<recipe_id>, and I ran into issues, e.g. needing to loop through the filtered recipe list to delete hidden recipes, and std::vector doesn't handle deletion well.

It's painlessly doable using both, except that `std::unordered_set` is a poor choice here (considerably small number of recipes + need for intersections). A `set` would be better.

> Then I did it like this and everything worked.

Simply making it work is not quite enough. The solution should also be reasonably good.
Use of `recipe_subset` here has certain drawbacks:

- Excessive memory consumption for absolutely no reason. These five `tree`s (3 `map`s + 2 embedded `set`s) are pure waste.
```c++
	std::map<const recipe *, int>
        std::map<std::string, std::set<const recipe *>>
        std::map<itype_id, std::set<const recipe *>>
```
  `npc`s would all have this too if it stayed a part of `player` (`tree`s still consume memory despite being empty);

- Wasted CPU cycles for managing this burden;
- More code to maintain (functions commented with "won't be needed");
- Undesirable dependency (since the class has a different purpose and can change for different reasons).

> I'll check it if works without mutable

It's not lazily initialized in a `const` function (like `learned_recipes` that you were mimicking is), so it would work.